sem: refactor `semConstExpr`

zerbina opened the pull request:
## Summary
- split the evaluation part into a separate procedure (`evalConstExpr`)
- use `nkError` nodes for reporting errors from `evalConstExpr`
- for a `const` definition, the expression is now not evaluated if
  its type is illegal in a `const` context

## Details
- `nkError` nodes are now prevented from reaching `vmgen`
- the `typeAllowed` check doesn't reject `ref` types for `const`s
  anymore
- disallowed non-nil values (such as for `ref`s) are now rejected via
  `ensureNilOnly` (which also looks at the *values*, not only at the
   *types*)
---
<!-- Note: section break (`---`) onwards is not in CI merge commit -->

## Notes for Reviewers
* the main motivation here is to prevent `nkError` nodes from reaching `vmgen`. If they do, they'd have to be passed through `mirgen`, which is not intended to be possible.

<!-- Pull Request(PR) Help

Before Merge Ensure:
* title reads like a short changelog line entry
* code includes tests and is documented
* leave the source better than before, but split out big reformats

See contributor (guide)[https://nim-works.github.io/nimskull/contributing.html]
for details, especially if you're new to this project.

Tips that make PRs easier:
* for big/impactful changes, start with chat/discussions to refine ideas
* refine the pull request message over time; don't have to nail it in one go
* handle the single commit message requirement at the end of review

saem commented on compiler/sem/sem.nim:
```diff
@@ -474,52 +478,67 @@ proc tryConstExpr(c: PContext, n: PNode): PNode =
+  # this happens when the overloadableEnums is enabled. We short-circuit
+  # evaluation in this case, as neither ``vmgen`` nor ``semfold`` know what to
+  # do with the sym-choice
+  # XXX: this doesn't seem like a good idea. Maybe report an error instead?
```
This silly bit is in const, let, var, initialization expression analysis, as well. I can't help but feel that should have been handled earlier. I just can't picture the solution quite yet. Something like symbolic type based evaluation should have been done which either resolves the choice or all this is in a higher level context (eg: generic), where ambiguity isn't an issue.

The levels thing is taking a bunch of my brain space and my mind is only trying to see a solution through that lens at the moment. 😬

saem commented on compiler/sem/semstmts.nim:
```diff
@@ -1080,14 +1080,26 @@ proc semNormalizedConst(c: PContext, n: PNode): PNode =
+  def = semExprWithType(c, def)
+  typ = typeAllowedOrError(typ, skConst, c, def, typFlags)
+
+  # XXX: ``fitNode`` is never used here. That doesn't seem correct, as the
```
I agree that looks to be an oversight, I think in many cases it'd to unnoticed, but it does seem like a bug.

zerbina replied:
This comment was an oversight by me. When splitting the evaluation part from `semConstExpr` and replacing its usage in `semNormalizedConst`, I also added the call to `semExprWithType` in order to mirror the behavior of `semConstExpr` also does.

Since a call to `fitNode` has to happen *after* a `semExprX` and there was none following the call to `semConstExpr`, I made the (incorrect) assumption that `fitNode` wasn't used. A few lines above however, the rhs is both analysed *and* fitted. In other words, there was no issue (except maybe that the already semantically analysed expression was analysed again).

saem commented on compiler/sem/semstmts.nim:
```diff
@@ -1080,14 +1080,26 @@ proc semNormalizedConst(c: PContext, n: PNode): PNode =
+      #      nodes are not considered when inlining constants. If there occurrs
+      #      an error during analysis of the ``const``, the resulting symbol
+      #      should likely be an ``skError`` symbol
+      localReport(c.config, e)
```
If we keep the error and not report it eagerly then we can propagate that into the productions below and register an skError.

zerbina replied:
Yep, but no `skError` symbol is produced right now. Since the logic in `semNormalizedConst` is fairly complex and this PR already does a good amount of other things besides refactoring `semConstExpr`, my thinking was that it'd be better to eagerly report the error for now.

saem commented on compiler/vm/vmcompilerserdes.nim:
```diff
@@ -40,7 +40,10 @@ import
 from compiler/ast/reports_sem import SemReport,
   reportAst
-from compiler/ast/report_enums import ReportKind
+from compiler/ast/report_enums import ReportKind, repVM
```
Damn, so close to removing it. 😅

I'll likely catch it in the rebase.

saem commented on compiler/vm/vmcompilerserdes.nim:
```diff
@@ -66,30 +69,36 @@ proc deserialize(c: TCtx, m: VmMemoryRegion, vt: PVmType, formal, t: PType, info
 proc deserializeObject(c: TCtx, m: VmMemoryRegion, vt: PVmType, f, con: PType, info: TLineInfo): PNode
 
-proc deserializeRef*(c: TCtx, slot: HeapSlotHandle, vt: PVmType; f, con: PType, info: TLineInfo): PNode =
+proc deserializeRef(c: TCtx, slot: HeapSlotHandle, vt: PVmType; f, con: PType, info: TLineInfo): PNode =
```
Yay, smaller surface

saem commented on compiler/vm/vmgen.nim:
```diff
@@ -2463,10 +2454,6 @@ proc gen(c: var TCtx; n: PNode; dest: var TDest; flags: TGenFlags = {}) =
   case n.kind
-  of nkError:
-    # XXX: do a better job with error generation
-    fail(n.info, vmGenDiagCannotGenerateCode, n)
```
Cool, these moved out.

saem reviewed (approved):
This is better, I think for the normalized const semantic analysis it should be possible to continue without reporting and produce an nk/SK error instead.

zerbina commented:
I reverted disallowing all non-nil `ref` values crossing the VM/compiler border again. Too many tests no longer pass then, and this PR is mainly meant as a refactoring. Instead, making sure that there are no non-nil `ref` values (non-nil closure environments are still allowed for compatibility) part of a `const` now happens in `semNormalizedConst`.

@saem If you're okay with the new changes, feel free to merge.

saem commented:
bors r+

This should be relatively easy to rebase... but I am putting a moritorum on big changes that bring in a lot of absolute hot garbage that is "Reports", until there is a strategy to get us out of this quagmire. I'll post more in the relevant PR and throw up a message in chat.

zerbina commented:
My intention here was not to make things harder than they already are, so I've converted this PR to a draft for now. The part that is a blocker for the MIR PR (`nkError` nodes reaching into `vmgen`) can be be worked around via an early-out in `evalConstExprAux`.

This way, moving to using an `skError` can also be implemented before later merging this PR, making the workaround that is currently employed obsolete.

zerbina committed:
sem: refactor `semConstExpr`

## Summary
- split the evaluation part into a separate procedure (`evalConstExpr`)
- use `nkError` nodes for reporting errors from `evalConstExpr`
- for a `const` definition, the expression is now not evaluated if
  its type is illegal in a `const` context

## Details
- `nkError` nodes are now prevented from reaching `vmgen`
- the `typeAllowed` check doesn't reject `ref` types for `const`s
  anymore
- disallowed non-nil values (such as for `ref`s) are now rejected via
  `ensureNilOnly` (which looks at the *values* not only the *types*)

### Misc
- fix reporting a `adVmUnsupportedNonNil` diagnostics crashing the
  compiler

zerbina commented:
I rebased the PR onto `devel` and adjusted it to the reporting-related changes plus to the `skError` usage in `semNormalizedConst`. The `localReport` that was previously used is now no longer necessary and thus removed.

bors r+

zerbina commented:
bors r+
